### DISCUSSION TITLE
Guidance on Replacing YOLOv8 Backbone with MobileNetV3 (Using Google Colab)

### DISCUSSION THREAD

**Discussion by eternalsummer01** [ultralytics/ultralytics#21130] [Q&A]
Hi everyone,

I'm planning to modify the YOLOv8 architecture to use MobileNetV3 as the backbone instead of the default one. I searched through previous issues and discussions but couldn't find much detailed information about this specific change.

So far, I understand that I’ll need to modify the .yaml configuration file in the repo to define the new backbone. I’ll be working on this in a Google Colab environment.

Could someone please clarify:
What are the exact steps to properly replace the backbone?
Do I need to write a custom MobileNetV3 module in PyTorch, or is there an existing one compatible with YOLOv8’s structure?
Are there any additional changes needed in the training or model loading scripts?
Is there any official or community guide/tutorial available on this?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @eternalsummer01, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will review your request and assist you soon! 😊

**Reply by glenn-jocher** [reply]
Hi @eternalsummer01! For replacing the YOLOv8 backbone with MobileNetV3, you'll need to create a custom model configuration. Here's the approach:

You'll need to define a custom YAML configuration that specifies MobileNetV3 as the backbone - this typically involves implementing the MobileNetV3 architecture in the model definition since it's not natively supported. The backbone replacement requires modifying both the model architecture and ensuring proper feature map outputs that match YOLO's neck expectations.

For the implementation, you can reference MobileNetV3 from torchvision and adapt it to work with YOLO's architecture, ensuring the output channels and feature map sizes align properly with the neck and head components.

I'd recommend checking the [custom model training guide](https://docs.ultralytics.com/modes/train/) and looking at existing model configurations in the `ultralytics/cfg/models/` directory for reference on how architectures are defined. You may also want to explore the [model customization documentation](https://docs.ultralytics.com/help/contributing/) for insights into extending the framework.

This is a fairly advanced modification that requires understanding both architectures - feel free to share your progress or specific implementation challenges as you work through it!

**Comment by Y-T-G** [comment]
You can find MobileNet configs here: 

https://github.com/Y-T-G/community/tree/main/cfg/detect

**Reply by eternalsummer01** [reply]
Hi, I’m working with your mobilenet_v3_small-fpn.yaml model config from the community repo. I’d like to run inference on the COCO128 dataset or a similar multi-class dataset.

Could you please clarify the full steps to:

Adjust or configure the YAML file for multi-class detection (e.g., COCO128 with 80 classes) instead of nc: 1.

Load the model properly in Google Colab.

Run inference on test images and interpret the output classes.

**Reply by glenn-jocher** [reply]
Hi @eternalsummer01! To use the MobileNet config for multi-class detection with COCO128 (80 classes):

1. **Modify the YAML**: Change `nc: 1` to `nc: 80` in the mobilenet_v3_small-fpn.yaml file to match COCO's 80 classes.

2. **Load and train the model**:
```python
from ultralytics import YOLO

# Load model with custom config
model = YOLO('path/to/mobilenet_v3_small-fpn.yaml')
# Train on COCO128
model.train(data='coco128.yaml', epochs=100)
```

3. **Run inference**:
```python
# After training, or load pre-trained weights
results = model('path/to/image.jpg')
# View results
results[0].show()
# Access class names and predictions
print(results[0].names)  # Class names dictionary
```

The model will automatically handle the 80 COCO classes once you've updated the `nc` parameter and trained on the COCO128 dataset. The output will include bounding boxes, confidence scores, and class predictions for all 80 COCO classes.

**Reply by Y-T-G** [reply]
@eternalsummer01 The model isn't trained. It's a config that you use to train a new model. So running inference on COCO128 wouldn't return any detections if you don't train it first.
